framework/src/FitFractionCalculator.cpp

```diff
@@ -0,0 +1,73 @@
+#include "FitFractionCalculator.h"
+#include <memory>
+#include <algorithm>
+#include <cctype>
+#include <iomanip>
+#include "TTree.h"
+
+FitFractionCalculator::FitFractionCalculator(IPDF& pdf, PhaseSpaceBoundary& boundary) : storedBoundary(boundary)
+{
+	vector<std::string> doNotIntegrate = pdf.GetDoNotIntegrateList();
+	for(const auto& combination: storedBoundary.GetDiscreteCombinations())
+	{
+		int i = GetIndex(combination);
+		double TotalIntegral = pdf.GetPDFIntegrator()->Integral(combination, &storedBoundary);
+		for(const auto& ComponentName: pdf.PDFComponents())
+		{
+			ComponentRef thisRef( ComponentName, "dummyObservable" );
+			ComponentFitFractions[i][ComponentName] = pdf.GetPDFIntegrator()->NumericallyIntegrateDataPoint(combination, &storedBoundary, doNotIntegrate, &thisRef);
+			ComponentFitFractions[i][ComponentName] /= TotalIntegral;
+		}
+	}
+}
+std::map<std::string, double> FitFractionCalculator::GetFitFractions(DataPoint* combination) const
+{
+	return ComponentFitFractions.at(GetIndex(combination));
+}
+double FitFractionCalculator::GetComponentFitFraction(DataPoint* combination, std::string ComponentName) const
+{
+	return ComponentFitFractions.at(GetIndex(combination)).at(ComponentName);
+}
+void FitFractionCalculator::Print() const
+{
+	for(const auto& combination: ComponentFitFractions)
+	{
+		std::cout << "Fit fractions in combination " << combination.first << "\n";
+		for(const auto& fraction: combination.second)
+			std::cout << std::setw(30) << fraction.first << " : " << fraction.second << "\n";
+		std::cout << std::endl;
+	}
+}
+void FitFractionCalculator::WriteToFile(std::string filename)
+{
+	auto OutputFile = std::unique_ptr<TFile>(TFile::Open(filename.c_str(),"UPDATE"));
```

Comment:
The documentation for `TFile::Open` says "the caller is responsible for deleting the pointer", so either I use a smart pointer or write a `delete` statement at the end of the scope. Given the choice of those two, I'd rather go with the one that doesn't involve manual memory management.